docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+    - Otherwise, `Ti` is fixed to be `Pi`, where `Pi` is the constraint solution
+      for `Ti` with respect to `C + Ti <: Ai`.
+
+### Upwards generic instantiation resolution
+
+Upwards resolution is the process by which the parameter types of a generic
+method (or constructor, etc) are matched against the actual argument types from
+an invocation of a method to find a solution for the missing type arguments that
+have not been fixed by downwards resolution.
+
+`[T0 -> M0, ..., Tn -> Mn]` is the upwards solution for an invocation of a
+generic method of type `<T0 extends B0, ..., Tn extends Bn>(P0, ..., Pk) -> Q`
+given actual argument types `R0, ..., Rk`, a partial solution `[T0 -> P0, ...,
+Tn -> Pn]` and a partial constraint set `Cp`:
+  - If `Ri <: Pi [T0, ..., Tn] -> Ci`
+  - And the full constraint resolution of `Cp + C0 + ... + Cn` for `<T0 extends
+B0, ..., Tn extends Bn>` given the initial partial solution `[T0 -> P0, ..., Tn
+-> Pn]` is `[T0 -> M0, ..., Tn -> Mn]`
+
+### Discussion
+
+The incorporation of the type bounds information is asymmetric and procedural:
+it iterates through the bounds in order (`Bi[R0/T0, ..., ?/Ti, ..., ?/Tn]`).  Is
+there a better formulation of this that is symmetric but still allows some
+propagation?
+
+### Inference rules
+
+- The expression `e as T` is inferred as `m as T` of type `T` in context `K`:
+  - If `e` is inferred as `m` in an empty context
+- The expression `x = e` is inferred as `x = m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `x ??= e` is inferred as `x ??= m` of type `UP(T, M)` in
+  context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `await e` is inferred as `await m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `J` where:
+    - `J` is `FutureOr<K>` if `K` is not `_`, and is `_` otherwise
+- The expression `e0 ?? e1` is inferred as `m0 ?? m1` of type `T` in
+  context `K`:
+  - If `e0` is inferred as `m0` of type `T0` in context `K`
+  - And `e1` is inferred as `m1` of type `T1` in context `J`
+  - Where `J` is `T0` if `K` is `_` and otherwise `K`
+  - Where `T` is the greatest closure of `K` with respect to `?` if `K` is not
+    `_` and otherwise `UP(T0, T1)`
+- The expression `e0..e1` is inferred as `m0..m1` of type `T` in context `K`
+  - If `e0` is inferred as `m0` of type `T` in context `K`
+  - And `e1` is inferred as `m1` of type `P` in context `_`
+- The expression `e0 ? e1 : e2` is inferred as `m0 ? m1 : m2` of type `T` in
+  context `K`
+  - If `e0` is inferred as `m0` of any type in context `bool`
+  - And `e1` is inferred as `m1` of type `T0` in context `K`
+  - And `e2` is inferred as `m2` of type `T1` in context `K`
+  - Where `T` is the greatest closure of `K` with respect to `?` if `K` is not
+    `_` and otherwise `UP(T0, T1)`
```

Comment:
The current implementation of type inference in analyzer doesn't follow this rule.  It always uses `UP(T0, T1)` for `T`.

I think analyzer's behavior is more intuitive.  Consider, for example:

    class C<T> {
      C(List<T> x);
    }

    main() {
      bool b = false;
      List<int> l1 = [1];
      List<int> l2 = [2];
      var x = new C(l1);
      var y = new C(l2);
      var z = new C(b ? l1 : l2);
    }

With analyzer's rule, we infer a type of `C<int>` for `x`, `y`, and `z`, which matches my intuitive expectation.  With the rule you've specified here, `z` gets a type of `C<dynamic>`, which IMHO is unexpected.